kotlinx-coroutines-core/common/src/selects/Select.kt

```diff
@@ -52,606 +97,696 @@ public interface SelectBuilder<in R> {
+        ClauseData<R>(clauseObject, regFunc, processResFunc, PARAM_CLAUSE_1, block, onCancellationConstructor).register()
+    override fun <P, Q> SelectClause2<P, Q>.invoke(param: P, block: suspend (Q) -> R) =
+        ClauseData<R>(clauseObject, regFunc, processResFunc, param, block, onCancellationConstructor).register()
 
-    private inline fun doResume(value: () -> Any?, block: () -> Unit) {
-        assert { isSelected } // "Must be selected first"
-        _result.loop { result ->
-            when {
-                result === UNDECIDED -> {
-                    val update = value()
-                    if (_result.compareAndSet(UNDECIDED, update)) return
-                }
-                result === COROUTINE_SUSPENDED -> if (_result.compareAndSet(COROUTINE_SUSPENDED, RESUMED)) {
-                    block()
+    /**
+     * Attempts to register this `select` clause. If another clause is already selected,
+     * this function does nothing and completes immediately.
+     * Otherwise, it registers this `select` instance in
+     * the [clause object][ClauseData.clauseObject]
+     * according to the provided [registration function][ClauseData.regFunc].
+     * On success, this `select` instance is stored as a waiter
+     * in the clause object -- the algorithm also stores
+     * the provided via [disposeOnCompletion] completion action
+     * and adds the clause to the list of registered one.
+     * In case of registration failure, the internal result
+     * (not processed by [ProcessResultFunction] yet) must be
+     * provided via [selectInRegistrationPhase] -- the algorithm
+     * updates the state to this clause reference.
+     */
+    protected fun ClauseData<R>.register(reregister: Boolean = false) {
+        // Is there already selected clause?
+        if (state.value.let { it is ClauseData<*> || it is Cancelled }) return
+        // For new clauses, check that there does not exist
+        // another clause with the same object.
+        if (!reregister) checkClauseObject(clauseObject)
+        // Try to register in the corresponding object.
+        if (tryRegister(this@SelectImplementation)) {
+            // Successfully registered, and this `select` instance
+            // is stored as a waiter. Add this clause to the list
+            // of registered clauses and store the provided via
+            // [invokeOnCompletion] completion action into the clause.
+            //
+            // Importantly, the [waitUntilSelected] function is implemented
+            // carefully to ensure that the cancellation handler has not been
+            // installed when clauses re-register, so the logic below cannot
+            // be invoked concurrently with the clean-up procedure.
+            if (!reregister) clauses!! += this
+            disposableHandle = this@SelectImplementation.disposableHandle
+            this@SelectImplementation.disposableHandle = null
```

Comment:
That's a great idea, thanks! I will do the refactoring.

Reply:
Actually, this change seems almost impossible, as I will add a special `disposeOnCompletion(segment, index)` method later, so no `DisposableHandle` instance will be returned in new channels and semaphore. Even now, semaphore modification seems weird with such API.

Reply:
I do not know how to make this part better. Yet, one may think that `selectInRegistrationPhase` should be invoked when `resume` is called for plain continuations (we almost always invoke it during registration). Similarly, `trySelect` should be invoked in the cases when `tryResume` Is called for continuations. Similarly, `disposeOnCompletion` is invoked instead of `invokeOnCompletion`.

Reply:
>Actually, this change seems almost impossible, as I will add a special disposeOnCompletion(segment, index) method later, so no DisposableHandle instance will be returned in new channels and semaphore. Even now, semaphore modification seems weird with such API.

Could you please elaborate on how exactly `disposableHandle` will be propagated to `ClauseData` if there is no disposable handle at all? I'm ready to take a look at any draft PR to get an idea of that, because for now it looks like this spilling protocol can be avoided

Reply:
There will be an addition `disposeOnCompletion(segment: Segment<*>, index: Int)` function to avoid creating `DisposableHandle`-s. I suggested the same optimization for continuations in #3084.

Reply:
But `DisposableHandle` protocol will still be required for things like `onTimeout`, won't it?

Reply:
With the optimization that avoids creating `DisposableHandle`-s, there will be two different APIs: one that manipulates `DisposableHandle`-s and one that uses `Segment` and `index` directly. That is why I am suggesting keeping the API as-is. Also, I do like the fact that the current API is consistent with the `CancellableContinuation` API, which makes it easier to support `select` further.